x-pack/platform/plugins/shared/fleet/server/services/epm/packages/update_custom_integration.ts

```diff
@@ -0,0 +1,181 @@
+
+import { getInstalledPackageWithAssets } from './get';
+
+import { installPackageWithStateMachine } from './install';
+
+export async function updateCustomIntegration(
+  esClient: ElasticsearchClient,
+  soClient: SavedObjectsClientContract,
+  id: string,
+  fields: {
+    readMeData?: string;
+    categories?: string[];
+  }
+) {
+  // Get the current integration using the id
+  let integration: SavedObject<IntegrationAttributes> | null = null;
+
+  try {
+    integration = await soClient.get(PACKAGES_SAVED_OBJECT_TYPE, id);
+  } catch (error) {
+    // Ignore the error and handle the case where integration is null later
+  }
+
+  // if theres no integration, the soClient will throw the error above and we will ignore it so we can handle it with our own here
+  if (!integration) {
+    throw new CustomIntegrationNotFoundError(`Integration with ID ${id} not found`);
+  } else if (
+    integration.attributes.install_source !== 'custom' &&
+    integration.attributes.install_source !== 'upload'
+  ) {
+    throw new NotACustomIntegrationError(`Integration with ID ${id} is not a custom integration`);
+  }
+
+  // add one to the patch version in the semver
+  const newVersion = integration.attributes.version.split('.');
+  newVersion[2] = (parseInt(newVersion[2], 10) + 1).toString();
+  const newVersionString = newVersion.join('.');
+  // Increment the version of everything and create a new package
+  const res = await incrementVersionAndUpdate(soClient, esClient, id, {
+    version: newVersionString,
+    readme: fields.readMeData,
+  });
+  return {
+    version: newVersionString,
+    status: res.status,
+  };
+}
+// Increments the version of everything, then creates a new package with the new version, readme, etc.
+export async function incrementVersionAndUpdate(
+  soClient: SavedObjectsClientContract,
+  esClient: ElasticsearchClient,
+  pkgName: string,
+  data: {
+    version: string;
+    readme: string | undefined;
+  }
+) {
+  const installedPkg = await getInstalledPackageWithAssets({
+    savedObjectsClient: soClient,
+    pkgName,
+  });
+
+  const assetsMap = [...installedPkg!.assetsMap.entries()].reduce((acc, [path, content]) => {
+    if (path === `${pkgName}-${installedPkg!.installation.install_version}/manifest.yml`) {
+      const yaml = load(content!.toString());
+      yaml.version = data.version;
+
+      content = Buffer.from(dump(yaml));
+    }
+
+    acc.set(
+      path.replace(`-${installedPkg!.installation.install_version}`, `-${data.version}`),
+      content
+    );
+    return acc;
+  }, new Map<string, Buffer | undefined>());
+
+  assetsMap.set(
+    `${pkgName}-${data.version}/docs/README.md`,
+    data.readme ? Buffer.from(data.readme) : undefined
+  );
+
+  // update the changelog asset as well by adding an entry
+  const changelogPath = `${pkgName}-${data.version}/changelog.yml`;
+  const changelog = assetsMap.get(changelogPath);
+  if (changelog) {
```

Comment:
Should we add a changelog item if there is none present? 
I tried to update the readme of a custom integration without one, created with 
```
POST kbn:/api/fleet/epm/custom_integrations
{
 "integrationName": "test",
 "datasets": []
}
```
and got the following error
<img width="1507" alt="Screenshot 2025-04-23 at 8 54 42 AM" src="https://github.com/user-attachments/assets/25f67afc-1803-4b47-bca8-9cddbd118472" />